README.md
# gvttc-ical
Tools to build iCalendar entries for GVTTC leagues

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar gvttc-ical-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-08-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-08-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/gvttc-ical/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/gvttc-ical/compare/0.1.0...0.1.1


test/gvttc_ical/core_test.clj
(ns gvttc-ical.core-test
  (:require [clojure.test :refer :all]
            [gvttc-ical.core :refer :all]
            [gvttc-ical.data-2015 :as data]
            ))

;; move to core


(deftest data-is-available
 (testing "data is available"
    (is (resolve 'data/players))
   (is (resolve 'data/monday-league))
  ))

(deftest test-plays-in?
  (testing "plays-in? predicate"
    (is (plays-in? :rprescott data/monday-league))
    (is (plays-in? :rprescott data/tuesday-league))
    (is (not (plays-in? :rprescott data/wednesday-league)))
    (is (= [:monday :tuesday] (map :type (filter #(plays-in? :rprescott %) data/leagues))))
))

;;(deftest test-monday-events
;;  (testing 
;;    (is 
   
(deftest monday-only-player
  (testing "a player from Monday only"
	  (let []
      (is :todo))))


;; @TODO: remove this
(run-tests)


test/gvttc_ical/data_2016.clj
(ns gvttc-ical.data-2016)

;; 2016 schedules online
;; Monday: https://drive.google.com/file/d/0B8etdrrcPnjHcjFCeDlSQ2wxQzA/view
;; Tuesday: https://drive.google.com/file/d/0B8etdrrcPnjHcjFCeDlSQ2wxQzA/view
;; Wednesday: https://drive.google.com/file/d/0B8etdrrcPnjHcjFCeDlSQ2wxQzA/view

(def players 
  {
   :rprescott {:name "Ralph Prescott" }
   :rmack {:name "Ray Mack" }
   ;; ...
})

(def monday-league  {
	 :dates (sorted-map ;; not sure if a sorted map is actually is needed
     #inst "2015-10-06" { :matches [[4 3] [8 10] [11 7] [1 6] [5 2] [9 12] ] }
     #inst "2015-10-05" { :matches [[] [] [] [] [] [] ] }
     #inst "2015-10-12" { :matches [[] [] [] [] [] [] ] }
     #inst "2015-10-19" { :matches [[] [] [] [] [] [] ] }
     #inst "2015-11-02" { :matches [[] [] [] [] [] [] ] }
     #inst "2015-11-09" { :matches [[] [] [] [] [] [] ] }
     #inst "2015-11-16" { :matches [[] [] [] [] [] [] ] }
     #inst "2015-11-23" { :matches [[] [] [] [] [] [] ] }
     #inst "2015-11-30" { :matches [[] [] [] [] [] [] ] }
     #inst "2015-12-07" { :matches [[] [] [] [] [] [] ] }
     #inst "2015-12-14" { :matches [[] [] [] [] [] [] ] }
     #inst "2015-12-21" { :type :off :text "Merry Christmas"}
	 )
;; ...
           
	 :teams 
    [
     nil ;; team 0
      [:amaronian :dfreeeman :ebizari :gmaronian] ;; team 1
			[] ;; team 2
			[] ;; team 3
			[] ;; team 4
			[] ;; team 5
			[:rprescott :tmchang :hshi :ddwight] ;; team 6
			[] ;; team 7
			[] ;; team 8
			[] ;; team 9
			[] ;; team 10
			[] ;; team 11
		]
   :tables 
   [
		"1 & Half of 2"
		"3 & Half of 2"
		"4 & Half of 5"
		"6 & Half of 5"
		"7 & 8"
    :bye
	 ]
})
   

src/gvttc_ical/core.clj
(ns gvttc-ical.core
	(:require 
    [clojure.tools.cli :refer [parse-opts]]
		[gvttc-ical.data-2015 :as data] ;; get dynamic loading to work
  )
  (:gen-class))

(defn plays-in? [pid league]
  (some #{pid} (flatten (:teams league))))

(defn all-players [all-leagues]
  (->> (map :teams data/leagues) flatten distinct (filter #(not (nil? %)))))

(defn event [p league night]
  (println night)
  (let
    [start-time (fn [d] (println "d: " (class d)) (. d setTime (:start-time league)))]
    { ;;:start-time (start-time (key night))
      :format :todo
      :vs-team :todo
      :vs-players [:todo :todo :todo :todo]
    }
  ))

(defn events [p league]
  (for [e (:events league)]
    (event p league e)))

(defn calendar-data [p leagues]
  {
    :player p
    :events ;; all events from all leagues player is in
       (for [l leagues
             :when (plays-in? (:id p) l)]
         (events p l))})

(defn data->ical [d] 
  d) ;; TODO: really implement this

(defn calendar [p leagues]
  (-> 
    (calendar-data p leagues) ;; calcuate the actual data
    (data->ical))) ;; convert data to icalendar format

(defn pid->player [pid players]
  (conj { :id pid } (pid players)))

;; NB: this function (and everything it calls) is unit testable
(defn all-calendars 
  "create a lazy sequence of the calendar for each player"
  [players leagues]
  (for [pid (all-players leagues)]
		(calendar (pid->player pid players) leagues)))

(defn cal->file [dir calendar]
  "file name to use for this calendar"
	(let [pid (get-in calendar [:player :id])]
	  (format "%s/%s.ics" dir (name pid))))

(defn output-all-calendars! [dir players leagues]
	(doseq 
		[calendar (all-calendars players leagues)
     file (cal->file dir calendar)]
        (println file "=" calendar)
        ;; something like: (spit file calendar)
        ))

;; TODO: integrate better CLI options using: https://github.com/clojure/tools.cli
(defn -main
  ""
  [& args]
  (let [year "2015" ;; TODO: parse from arguments
        data (str  "data-" year) 
        dir "ical" ] 
  (println data)
  ;; (require '(gvttc-ical data-2015)) ;; TODO: get dynamic loading to work
  (output-all-calendars! dir data/players data/leagues)
  ))

src/gvttc_ical/data_2015.clj
(ns gvttc-ical.data-2015)
;; TODO: consider switching to using EDN

;; 2016 schedules online
;; Monday: https://drive.google.com/file/d/0B8etdrrcPnjHcjFCeDlSQ2wxQzA/view
;; Tuesday: https://drive.google.com/file/d/0B8etdrrcPnjHcjFCeDlSQ2wxQzA/view
;; Wednesday: https://drive.google.com/file/d/0B8etdrrcPnjHcjFCeDlSQ2wxQzA/view

(def players 
  {
   :amaronian {:name "Andre Maronian" }
   :rprescott {:name "Ralph Prescott" }
   :rmack {:name "Ray Mack" }
   ;; ...
})

(def monday-league  {
   :type :monday
   :start-time 18 ;; 6PM
   :end-time 22 ;; 10PM
	 :events (sorted-map ;; not sure if a sorted map is actually is needed
		#inst "2015-10-06" { :matches [[4 3] [8 10] [11 7] [1 6] [5 2] [9 12] ] }
    #inst "2015-10-05" { :matches [[] [] [] [] [] [] ] }
		;;#inst "2015-10-12" { :matches [[] [] [] [] [] [] ] }
		;;#inst "2015-10-19" { :matches [[] [] [] [] [] [] ] }
		;;#inst "2015-11-02" { :matches [[] [] [] [] [] [] ] }
		;;#inst "2015-11-09" { :matches [[] [] [] [] [] [] ] }
		;;#inst "2015-11-16" { :matches [[] [] [] [] [] [] ] }
		;;#inst "2015-11-23" { :matches [[] [] [] [] [] [] ] }
		;;#inst "2015-11-30" { :matches [[] [] [] [] [] [] ] }
		;;#inst "2015-12-07" { :matches [[] [] [] [] [] [] ] }
		;;#inst "2015-12-14" { :matches [[] [] [] [] [] [] ] }
		;;#inst "2015-12-21" { :type :off :text "Merry Christmas"}
	 )
;; ...
           
	 :teams 
   [
     nil ;; team 0
     [:amaronian :dfreeeman :ebizari :gmaronian] ;; team 1
		 [] ;; team 2
		 [] ;; team 3
		 [] ;; team 4
		 [] ;; team 5
		 [:rprescott :tmchang :hshi :ddwight] ;; team 6
		 [] ;; team 7
		 [] ;; team 8
		 [] ;; team 9
		 [] ;; team 10
		 [] ;; team 11
	 ]
   :tables 
   [
		"1 & Half of 2"
		"3 & Half of 2"
		"4 & Half of 5"
		"6 & Half of 5"
		"7 & 8"
    :bye
	 ]
})

(def tuesday-league  
  {
   :type :tuesday
   :teams 
   [
     nil ;; team 0
     [] ;; team 1
		 [] ;; team 2
		 [:pyu :rprescott :bstein] ;; team 3
   ]
})

;; https://drive.google.com/file/d/0B8etdrrcPnjHcjFCeDlSQ2wxQzA/view
(def wednesday-league 
  {
   :type :wednesday })

(def leagues [monday-league tuesday-league wednesday-league])
   